File: lib/markov-chain.rb
#-*- coding: utf-8 -*-

require "natto"

module NukokusaBot

  class Natto::MeCabNode
    def compress
      self.feature[0, 2] + self.surface
    end
  end

  class MarkovChain

    def initialize
      @word_id  = Hash.new  # {id1 => 'word1', id2 => 'word2'}
      @tailings = Hash.new  # {id1 => {id2 => count, id3 => count}}
      @mecab = Natto::MeCab.new

      @max_id  = -1
      register_words(['末尾', '先頭'])
    end

    def add(sentence)
      words = @mecab.parse_as_nodes(sentence)[0..-2].map(&:compress)

      register_words(words)

      count_tailings('先頭', words.first)
      words.each_cons(2){ |pre, suc| count_tailings(pre, suc) }
      count_tailings(words.last, '末尾')

      require "pp"
      pp @tailings
    end
    alias_method :<<, :add
    alias_method :append, :add

    def generate(limit = 140)
      puts
      chain_ids = [rand(2..@max_id)]

      chain_ids << choose_next_of(chain_ids.last) until chain_ids.last == 0

      chain_to_string(chain_ids)
    end

    private

    def choose_next_of(id)
      candidates = Array.new
      @tailings[id].each do |id, count|
        candidates << [id] * count
      end
      candidates.flatten.sample
    end

    def chain_to_string(ids)
      puts @word_id.invert.values_at(*ids).collect{|word| word[2..-1]}.join
    end

    def register_words(words)
      words.each do |word|
        unless @word_id.has_key? word
          @max_id += 1
          @word_id[word] = @max_id
        end
      end
    end

    def count_tailings(pre, suc)
      pre_id = @word_id[pre]
      suc_id = @word_id[suc]

      if @tailings.has_key? pre_id
        if @tailings[pre_id].has_key? suc_id
          @tailings[pre_id][suc_id] += 1
        else
          @tailings[pre_id][suc_id] = 1
        end
      else
        @tailings[pre_id] = Hash.new
        @tailings[pre_id][suc_id] = 1
      end
    end
  end
end

File: src/main/scala/MarkovChain.scala
package net.joe_noh.nukokusabot

import twitter4j._
import scala.util.Random
import scala.collection.JavaConversions._

import net.reduls.gomoku._

import com.github.aselab.activerecord._
import com.github.aselab.activerecord.dsl._

class MarkovChain {

  val random = new Random()

  def addStatus(status: Status): Unit = {
    if (!Tweet.exists(_.tweetId === status.getId())) {
      Tweet(status.getId(), status.getText()).save()
      updateDictionary(status.getText())
    }
  }

  def addStatus(id: Long, text: String): Unit = {
    if (!Tweet.exists(_.tweetId === id)) {
      Tweet(id, text).save()
      updateDictionary(text)
    }
  }

  def rebuildDictionary = {
    Word.deleteAll
    Link.deleteAll
    (0 to (Tweet.count/100).toInt).foreach { i =>
      Tweet.page(100*i, 100).foreach { status  =>
        addStatus(status.id, status.text)
      }
    }
  }

  def generateSentence(limit: Int = 140): String = {
    def challenge: String = {
      def buildChain(preId: Long, idAcc: List[Long]): List[Long] = {
        val nextId = fetchNextId(preId)
        random.nextDouble match {
          case d: Double if d < 0.1 => idAcc
          case _: Double  => buildChain(nextId, idAcc :+ nextId)
        }
      }

      val id = random.nextLong() % Word.count
      val sentence = translateChainToJapanese(buildChain(id, List(id)))
      sentence match {
      	case _ if (sentence.length == 0 || sentence.length > limit) => challenge
        case _ => sentence
      }
    }
    challenge
  }

  private def translateChainToJapanese(chain: List[Long]) = {
    def translate(chain: List[Long], stringAcc: String): String = chain match {
      case head::rest => translate(rest, stringAcc + Word.find(head).get.surface)
      case _ => stringAcc
    }
    translate(chain, "")
  }


  private def updateDictionary(sentence: String): Unit = {
    val pairList = Tagger.parse(sentence).map(m => (m.surface, m.feature)).toList

    val words = pairList.map{ p =>
      val feature = Feature.findByOrCreate(Feature(p._2), "tag")
      Word.findByOrCreate(Word(p._1, feature.id.asInstanceOf[Int]), "surface", "featureId")
    }

    words.sliding(2).foreach{ case List(pre, suc) =>
      val link = Link.findByOrCreate(Link(pre.id, suc.id), "preId", "sucId")
      link.count += 1
      link.save()
    }
  }

  private def fetchNextId(wordId: Long): Long = {
    var candidates = Link.findBy("preId", wordId).toList.collect{ case link => List.fill(link.count)(link.sucId) }.flatten
    random.shuffle(candidates).get(0)
  }
}
